src/base/emu-i386/simx86/interp.c

```diff
@@ -1789,30 +1789,38 @@ intop3b:		{ int op = ArOpsFR[D_MO(opc)];
 			temp=0; POP(mode, &temp);
 			if (CONFIG_CPUSIM) RFL.valid = V_INVALID;
 			if (V86MODE()) {
+			    int is_tf;
 stack_return_from_vm86:
 			    if (debug_level('e')>1)
 				e_printf("Popped flags %08x fl=%08x\n",
 					temp,EFLAGS);
+			    is_tf = !!(EFLAGS & TF);
 			    if (IOPL==3) {	/* Intel manual */
+				/* keep reserved bits + IOPL,VIP,VIF,VM,RF */
 				if (mode & DATA16)
-				    FLAGS = temp;	/* oh,really? */
+				    FLAGS &= ~(SAFE_MASK|EFLAGS_IF);
 				else
-				    EFLAGS = (temp&~0x1b3000) | (EFLAGS&0x1b3000);
+				    EFLAGS &= ~(SAFE_MASK|EFLAGS_IF);
+				EFLAGS |= temp & (SAFE_MASK|EFLAGS_IF);
 			    }
 			    else {
-				int is_tf = !!(EFLAGS & TF);
 				/* virtual-8086 monitor */
 				/* move mask from pop{e}flags to regs->eflags */
-				EFLAGS = (EFLAGS & ~SAFE_MASK) | (temp & SAFE_MASK);
-				if (temp & EFLAGS_IF) {
+				if (mode & DATA16)
+				    FLAGS &= ~SAFE_MASK;
+				else
+				    EFLAGS &= ~SAFE_MASK;
+				EFLAGS |= temp & SAFE_MASK;
+				if (temp & EFLAGS_IF)
 				    EFLAGS |= EFLAGS_VIF;
-				    if (vm86s.regs.eflags & VIP) {
-					if (debug_level('e')>1)
-					    e_printf("Return for STI fl=%08x\n",
-						EFLAGS);
-					TheCPU.err = (is_tf ? EXCP01_SSTP : EXCP_STISIGNAL);
-					return PC + (opc==POPF);
-				    }
+			    }
+			    if (temp & EFLAGS_IF) {
+				if (vm86s.regs.eflags & VIP) {
```

Comment:
Grr, but I think VIP doesn't work
in IOPL=3 mode, or does it?
But I see that you check it completely
from an outside (from vm86s), and I
suppose its the simplest possible cheat.
So I guess lets apply that and re-impl
VME/PVI later.

Reply:
yes it's from the outside on purpose, much like the `request_interrupt_window` flag in KVM.

the trouble with PVI of course is that it doesn't affect pushf/popf, so I see no extra value in using that (unless IOPL=3 turns out to be problematic)